collect_app/src/main/java/org/odk/collect/android/activities/viewmodels/FormMapViewModel.java

```diff
@@ -0,0 +1,194 @@
+package org.odk.collect.android.activities.viewmodels;
+
+import androidx.lifecycle.ViewModel;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.odk.collect.android.R;
+import org.odk.collect.android.forms.Form;
+import org.odk.collect.android.geo.MapFragment;
+import org.odk.collect.android.geo.MapPoint;
+import org.odk.collect.android.instances.Instance;
+import org.odk.collect.android.instances.InstancesRepository;
+import org.odk.collect.android.provider.InstanceProviderAPI;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.Nullable;
+
+import timber.log.Timber;
+
+public class FormMapViewModel extends ViewModel {
+    /**
+     * The form that is mapped.
+     */
+    private final Form form;
+
+    /**
+     * Cached count of all of this form's instances, including ones that don't have a geometry to
+     * map.
+     */
+    private int totalInstanceCount;
+
+    /**
+     * Quick lookup of instance objects from map feature IDs.
+     */
+    private final Map<Integer, Instance> instancesByFeatureId = new HashMap<>();
+
+    /**
+     * Points to be mapped. Note: kept separately from {@link #instancesByFeatureId} so we can
+     * quickly zoom to bounding box.
+     */
+    private final List<MapPoint> points = new ArrayList<>();
+
+    /**
+     * True if the map viewport has been initialized, false otherwise.
+     */
+    private boolean viewportInitialized;
+
+    private final InstancesRepository instancesRepository;
+
+    public FormMapViewModel(Form form, InstancesRepository instancesRepository) {
+        this.instancesRepository = instancesRepository;
+        this.form = form;
+    }
+
+    public String getFormTitle() {
+        return form.getDisplayName();
+    }
+
+    public String getFormId() {
+        return form.getJrFormId();
+    }
+
+    public int getInstanceCount() {
+        return totalInstanceCount;
+    }
+
+    public int getMappedPointCount() {
+        return points.size();
+    }
+
+    /**
+     * Clears the existing features on the given map and places features for the current form's
+     * instances. If features were added to the map and the map view was not previously initialized,
+     * zooms to display all points.
+     */
+    public void mapUpdateRequested(MapFragment map) {
```

Comment:
I think the ViewModel is serving as a wrapper for a bunch of logic concerned with manipulating `MapFragment`. I'm not sure why we can't have methods like this inside `FormMapActivity`. I really think the mapping between map "features" and instances needs to live somewhere else - it feels like a detail of the presentation implementation rather than "state". What I do like about the ViewModel is that it protects the `FormMapActivity` from where instances actually come from, the `Form` object and also the logic around which instances should end up on the map. For me, the ViewModel could be more like this:

```java
public class FormMapViewModel extends ViewModel {
    /**
     * The form that is mapped.
     */
    private final Form form;

    private final InstancesRepository instancesRepository;

    public FormMapViewModel(Form form, InstancesRepository instancesRepository) {
        this.instancesRepository = instancesRepository;
        this.form = form;
    }

    public String getFormTitle() {
        return form.getDisplayName();
    }

    public String getFormId() {
        return form.getJrFormId();
    }

    public List<GeoFormInstance> getInstances() {
        List<Instance> instances = instancesRepository.getAllBy(form.getJrFormId());
        List<GeoFormInstance> geoInstances = new ArrayList<>();

        for (Instance instance : instances) {
            if (instance.getGeometry() != null) {
                try {
                    JSONObject geometry = new JSONObject(instance.getGeometry());
                    switch (instance.getGeometryType()) {
                        case "Point":
                            JSONArray coordinates = geometry.getJSONArray("coordinates");
                            // In GeoJSON, longitude comes before latitude.
                            double lon = coordinates.getDouble(0);
                            double lat = coordinates.getDouble(1);
                            geoInstances.add(new GeoFormInstance(
                                    instance.getDatabaseId(),
                                    new Pair<>(lat, lon),
                                    getStatusForInstance(instance)
                            ));
                    }
                } catch (JSONException e) {
                    Timber.w("Invalid JSON in instances table: %s", instance.getGeometry());
                }
            }
        }


        return geoInstances;
    }

    private Status getStatusForInstance(Instance instance) {
        if (instance != null) {
            if (instance.getDeletedDate() != null) {
                return Status.DELETED;
            }

            if ((instance.getStatus().equals(InstanceProviderAPI.STATUS_COMPLETE)
                    || instance.getStatus().equals(InstanceProviderAPI.STATUS_SUBMITTED)
                    || instance.getStatus().equals(InstanceProviderAPI.STATUS_SUBMISSION_FAILED))
                    && !instance.canEditWhenComplete()) {
                return Status.NOT_VIEWABLE;
            } else if (instance.getDatabaseId() != null) {
                if (instance.getStatus().equals(InstanceProviderAPI.STATUS_SUBMITTED)) {
                    return Status.VIEW_ONLY;
                }
                return Status.EDITABLE;
            }
        }

        return Status.UNKNOWN;
    }

    public static class GeoFormInstance {

        private final Long instanceID;
        private final Pair<Double, Double> geometry;
        private final Status status;

        public GeoFormInstance(Long instanceID, Pair<Double, Double> geometry, Status status) {
            this.instanceID = instanceID;
            this.geometry = geometry;
            this.status = status;
        }

        public Long getInstanceID() {
            return instanceID;
        }

        public Pair<Double, Double> getGeometry() {
            return geometry;
        }

        public Status getStatus() {
            return status;
        }
    }

    public enum Status {
        DELETED, NOT_VIEWABLE, VIEW_ONLY, EDITABLE, UNKNOWN
    }
}
```

Here the ViewModel is just responsible for "presenting" its client with the data it needs to render. We don't need to create a new object representation for the `Instance` here but I like it as an option as it lets us remove any of the logic around mapping status from one the Instance cares about to one our Activity cares about. In this setup `mapUpdateRequested` would move to the Activity and would just update the features on the map based on the data it's presented:

```java
private void mapUpdateRequested(MapFragment map) {
   points.clear();
   map.clearFeatures();

   List<FormMapViewModel.GeoFormInstance> instances = viewModel.getInstances();
   totalInstanceCount = instances.size();
   for (FormMapViewModel.GeoFormInstance instance : instances) {
        MapPoint point = new MapPoint(instance.getGeometry().first, instance.getGeometry().second);
        int featureId = map.addMarker(point, false);

        int drawableId = getDrawableIdForStatus(instance.getStatus());
        map.setMarkerIcon(featureId, drawableId);

        instancesByFeatureId.put(featureId, instance);
    }

    if (!viewportInitialized && !points.isEmpty()) {
        mapZoomToBoundingBoxRequested(map);
        viewportInitialized = true;
    }
}
```

I think there is something missing from our `MapFragment` interface here that's making this all a little awkward. Maybe we need a way to give features some kind of "meta identifier" like `setTag` on Android `View`. I'm not 100% sure. 

Does any of this make sense or is it silly?

Reply:
Ah, shoot, I thought I was on board with this but it has the problem that I highlighted above -- the featureId/dbID map, points and viewportInitialized are lost on orientation change. Do you suggest handling saving those manually? I'd also like unit tests for viewport initialization and that's no longer possible.

I think it's unfortunate that Google called their lifecycle-aware container `ViewModel`. I think it really wants to be more of a kind of controller. The first sentence in its description is `The ViewModel class is designed to store and manage UI-related data in a lifecycle conscious way.` To me, `viewportInitialized`, `points`, is UI-related data.

Reply:
But you're rebuilding the featureId/dbID map every time in `onResume` anyway right? Why does that need to persist between rotations? I'm wondering if my comment above about `configChanges` is relevant here and I'm confused about what happens on rotation - I'm assuming `onResume` is called.

Yeah it's not a great distinction. I'd refer to it as "data the UI needs" and that `points` etc are "state about the view".

Reply:
> you're rebuilding the featureId/dbID map every time in onResume anyway right?

Sigh. That's too bad.